Sort modes and modalities

glittershark opened the pull request:
`x @ mode1 mode2` and `x @ mode2 mode1` are equivalent (right?), as are things like `value mod mode1 mode2` and `value mod mode2 mode1` (right?). It'll  be nice to have a canonical order on these to avoid thrashing and inconsistency in code with modalities and mode annotations - I decided to just go with sorting by String.compare for now, rather than trying to impose any opinions on the order.

glittershark committed:
Sort modes and modalities

x @ mode1 mode2 and x @ mode2 mode1 are equivalent, as are things like `value
mod mode1 mode2` and `value mod mode2 mode1`. It'll be nice to have a canonical
order on these to avoid thrashing and inconsistency in code with modalities and
mode annotations. I decided to just go with sorting by String.compare for now,
rather than trying to impose any opinions on the order.

Signed-off-by: Aspen Smith <aspsmith@janestreet.com>

goldfirere reviewed (approved):
I think putting these in a canonical order is good. But I'm a little worried that it would be better to order sorting by axis rather than by name of the mode... I think I would find it jarring to change a mode from `aliased global` to `global unique`. That said, we expect to have mode abbreviations at some point, which would make my axis-oriented sorting not work.

In any case, this is an improvement over the status quo, so I'm approving. If you think my axis-oriented approach might be better, give it a shot. If not, just go ahead and merge.

dvulakh committed:
normalize mode order in the extended parser

the current state of the PR leaves the extended parse tree
unnormalized, but this is hidden by the wonderful bug in the
round-trip check

this commit sorts the modes while parsing the extended ast as we are
wont; now we don't need to handle them specially from `Fmt_ast.ml` or
do a separate normalization

Signed-off-by: David Vulakh <dvulakh@janestreet.com>
